File: 20-actors/watatsuna/methods/kotoba.py
#!/usr/bin/env python3
"""kotoba.py — watatsuna kotoba Datom-log writer (local, content-addressed). ADR-2606012600
+ ADR-2605262130 + ADR-2605312345.

The substrate boundary (root CLAUDE.md): canonical state is the **kotoba Datom log** —
content-addressed EAVT assertions, append-only (非終末論). watatsuna had a graph + analyzer but no
self-driving loop and no local log; this module is the **local, autonomous-loop** write path — the
same path shionome / ipaddress / yabai / sukashi use (`methods/autorun.py`): a self-driving
heartbeat appends content-addressed transactions to a local append-only EDN log with NO external
I/O, so watatsuna can run its own observe→analyze→persist submarine-cable-resilience cycle on the
Murakumo fleet without a human or a live node in the loop.

Constitutional posture is preserved by construction (watatsuna hard rules): every derived signal
is a RESILIENCE map (chokepoint-load, station-degree, cable-diversity, redundancy-gap) — never a
"where to cut" / interdiction framing (G2); public-record cable data only (G1); fault kinds record
only the public bulletin's own classification, never a sabotage verdict (G4). The loop persists
exactly what `analyze.py` already computes, with derived signals flagged :resilience/derived.

  - graph_datoms(rows)            → EAVT assertions for every entity (cable / station / link /
                                     segment / fault). E = the entity's id; lists fan out.
  - derived_datoms(cables, st, a) → EAVT assertions for the analyzer's derived :resilience/*
                                     signals (chokepoint-load, station degree/capacity,
                                     cable-diversity, redundancy-gap), flagged :resilience/derived.
  - make_tx / append_tx / read_log / head_cid / verify_chain — content-addressed commit-DAG.

EAVT = [op entity attribute value]; op is :db/add only (append-only — no :db/retract). Stdlib only.
Deterministic: the caller supplies tx_id + as_of (no wall clock) → resume-safe.
"""
from __future__ import annotations

import hashlib
import json
import os
import pathlib
import sys
from typing import Any

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from analyze import _parse, _tokens  # noqa: E402  (the inline EDN reader lives in analyze.py)

LOG_DEFAULT = (pathlib.Path(__file__).resolve().parents[1] / "data"
               / "watatsuna.datoms.kotoba.edn")

ID_KEYS = (":cable/id", ":station/id", ":cable.link/id", ":cable.seg/id", ":cable.fault/id")


def _add(entity: str, attr: str, value: Any) -> list:
    """One append-only EAVT assertion: [:db/add <entity> <attr> <value>]."""
    return [":db/add", entity, attr, value]


def graph_datoms(rows: list) -> list[list]:
    """Flatten the submarine-cable graph into append-only EAVT assertions. E = the entity's id;
    cardinality-many list values (e.g. :station/chokepoint, :cable.seg/traverses) fan out."""
    out: list[list] = []
    for r in rows:
        if not isinstance(r, dict):
            continue
        e = next((r[k] for k in ID_KEYS if k in r), None)
        if e is None:
            continue
        for k, v in r.items():
            if k in ID_KEYS:
                continue
            for item in (v if isinstance(v, list) else [v]):
                out.append(_add(e, k, item))
    return out


def derived_datoms(cables: dict, stations: dict, a: dict) -> list[list]:
    """Flatten the analyzer's derived :resilience/* signals into EAVT assertions, each flagged
    :resilience/derived true (a RESILIENCE map recomputed on read, never re-ingested as fact, and
    never an interdiction target-list — G2). Mirrors analyze.render_datoms. `a` is analyze.analyze()."""
    out: list[list] = []
    for cp in sorted(a["choke_load"], key=lambda k: -a["choke_load"][k]):
        e = f"resilience-choke-{cp}"
        out += [_add(e, ":resilience/chokepoint", cp),
                _add(e, ":resilience/chokepoint-load", a["choke_load"][cp]),
                _add(e, ":resilience/cable-count", a["choke_count"][cp]),
                _add(e, ":resilience/derived", True)]
    for s in sorted(stations, key=lambda k: -a["station_degree"][k]):
        if a["station_degree"][s] == 0:
            continue
        e = f"resilience-station-{s}"
        out += [_add(e, ":resilience/station", s),
                _add(e, ":resilience/station-degree", a["station_degree"][s]),
                _add(e, ":resilience/station-capacity-tbps", a["station_capacity"][s]),
                _add(e, ":resilience/derived", True)]
    for c in sorted(cables, key=lambda k: a["cable_diversity"][k]):
        e = f"resilience-cable-{c}"
        out += [_add(e, ":resilience/cable", c),
                _add(e, ":resilience/cable-diversity", a["cable_diversity"][c]),
                _add(e, ":resilience/derived", True)]
    # redundancy-gap: landing stations served by a single cable (routed to watatsumi 敷設, never to cut)
    for s in a["redundancy_gap"]:
        e = f"resilience-gap-{s}"
        out += [_add(e, ":resilience/redundancy-gap-station", s),
                _add(e, ":resilience/station-degree", a["station_degree"][s]),
                _add(e, ":resilience/derived", True)]
    return out


def _canonical(datoms: list[list], prev_cid: str) -> bytes:
    return json.dumps({"prev": prev_cid, "datoms": datoms},
                      ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode("utf-8")


def tx_cid(datoms: list[list], prev_cid: str = "") -> str:
    """Content address = sha256 over (prev_cid, datoms) → a commit-DAG."""
    return "b" + hashlib.sha256(_canonical(datoms, prev_cid)).hexdigest()


def make_tx(datoms: list[list], *, tx_id: int, as_of: int, prev_cid: str = "") -> dict:
    return {
        ":tx/id": tx_id,
        ":tx/as-of": as_of,
        ":tx/prev": prev_cid,
        ":tx/cid": tx_cid(datoms, prev_cid),
        ":tx/count": len(datoms),
        ":tx/datoms": datoms,
    }


def _edn_val(v: Any) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return repr(v)
    if isinstance(v, str):
        return v if v.startswith(":") else json.dumps(v, ensure_ascii=False)
    if isinstance(v, list):
        return "[" + " ".join(_edn_val(x) for x in v) + "]"
    return json.dumps(str(v), ensure_ascii=False)


def _tx_to_edn(tx: dict) -> str:
    datoms = " ".join("[" + " ".join(_edn_val(x) for x in d) + "]" for d in tx[":tx/datoms"])
    return (f'{{:tx/id {tx[":tx/id"]} :tx/as-of {tx[":tx/as-of"]} '
            f':tx/prev {json.dumps(tx[":tx/prev"])} :tx/cid {json.dumps(tx[":tx/cid"])} '
            f':tx/count {tx[":tx/count"]} :tx/datoms [{datoms}]}}')


def append_tx(tx: dict, log_path: pathlib.Path = LOG_DEFAULT) -> str:
    """Append ONE transaction to the append-only log (never rewrites). Returns the tx CID."""
    log_path.parent.mkdir(parents=True, exist_ok=True)
    if not log_path.exists():
        log_path.write_text(";; watatsuna kotoba Datom log — append-only EAVT transactions "
                            "(content-addressed DAG). Resilience map, never interdiction. "
                            "DO NOT hand-edit. ADR-2606012600.\n", encoding="utf-8")
    with log_path.open("a", encoding="utf-8") as fh:
        fh.write(_tx_to_edn(tx) + "\n")
    return tx[":tx/cid"]


def read_log(log_path: pathlib.Path = LOG_DEFAULT) -> list[dict]:
    """Read the log back as a list of transaction dicts (uses analyze.py's inline EDN reader)."""
    if not log_path.exists():
        return []
    txs = []
    for line in log_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith(";"):
            continue
        txs.append(_parse(_tokens(line)))
    return txs


def head_cid(log_path: pathlib.Path = LOG_DEFAULT) -> str:
    txs = read_log(log_path)
    return txs[-1][":tx/cid"] if txs else ""


def verify_chain(log_path: pathlib.Path = LOG_DEFAULT) -> dict:
    """Recompute every CID from its datoms + prev; verify the DAG is intact. {ok, length, broken_at}."""
    txs = read_log(log_path)
    prev = ""
    for i, tx in enumerate(txs):
        expect = tx_cid(tx.get(":tx/datoms", []), prev)
        if tx.get(":tx/cid") != expect or tx.get(":tx/prev") != prev:
            return {"ok": False, "length": len(txs), "broken_at": i}
        prev = tx[":tx/cid"]
    return {"ok": True, "length": len(txs), "broken_at": -1}


if __name__ == "__main__":
    import pathlib as _pl
    from analyze import load_edn

    here = _pl.Path(__file__).resolve().parents[1]
    g = here / "data" / "cable-graph.merged.kotoba.edn"
    if not g.exists():
        g = here / "data" / "seed-cable-graph.kotoba.edn"
    datoms = graph_datoms(load_edn(g))
    tx = make_tx(datoms, tx_id=1, as_of=20260608, prev_cid="")
    print(f"# watatsuna kotoba Datom tx — {tx[':tx/count']} EAVT assertions, cid={tx[':tx/cid'][:18]}…")

File: 20-actors/watatsuna/methods/kotoba.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/kotoba.py — the local content-addressed Datom-log writer.
(ns watatsuna.methods.kotoba
  "kotoba.clj — watatsuna kotoba Datom-log writer (local, content-addressed).
  ADR-2606012600 + ADR-2605262130 + ADR-2605312345.

  Canonical state is the kotoba Datom log — content-addressed EAVT assertions, append-only
  (非終末論). This is the local, autonomous-loop write path (the shionome/ipaddress/yabai/sukashi
  shape): a heartbeat appends content-addressed transactions to a local append-only EDN log with
  NO external I/O, so watatsuna runs its observe→analyze→persist cable-resilience cycle on the
  Murakumo fleet with no human/live-node in the loop.

  Constitutional posture holds by construction: every derived signal is a RESILIENCE map
  (chokepoint-load / station-degree / cable-diversity / redundancy-gap) — never a 'where to cut'
  interdiction framing (G2); public-record cable data only (G1); fault kinds carry only the
  public bulletin's own classification (G4).

    graph-datoms   → EAVT assertions for every entity (cable/station/link/segment/fault)
    derived-datoms → EAVT assertions for the analyzer's derived :resilience/* signals
    make-tx / append-tx / read-log / head-cid / verify-chain — content-addressed commit-DAG

  EAVT = [op entity attribute value]; op is :db/add only (append-only — no :db/retract).
  Deterministic: the caller supplies tx-id + as-of (no wall clock) → resume-safe. The CID is a
  clj-native content address (sha256 over the canonical pr-str of {:datoms :prev}); it is
  internally consistent + tamper-evident (it does not reproduce kotoba.py's JSON-based bytes,
  since the clj port carries real EDN keywords where the Python port carried strings)."
  (:require [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn log-default []
  (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "watatsuna.datoms.kotoba.edn")))

(def id-keys #{:cable/id :station/id :cable.link/id :cable.seg/id :cable.fault/id})

(defn- add-datom [e a v] [:db/add e a v])

(defn graph-datoms
  "Flatten the cable graph into append-only EAVT assertions. E = the entity's id; list values
  (e.g. :station/chokepoint, :cable.seg/traverses) fan out."
  [rows]
  (vec (mapcat
        (fn [r]
          (when (map? r)
            (when-let [e (some #(get r %) id-keys)]
              (for [[k v] r :when (not (id-keys k))
                    item (if (sequential? v) v [v])]
                (add-datom e k item)))))
        rows)))

(defn derived-datoms
  "Flatten the analyzer's derived :resilience/* signals into EAVT assertions, each flagged
  :resilience/derived true (a RESILIENCE map recomputed on read, never re-ingested as fact,
  never an interdiction target-list — G2). `a` is analyze/analyze."
  [{:keys [cables stations]} a]
  (vec
   (concat
    (mapcat (fn [cp]
              (let [e (str "resilience-choke-" cp)]
                [(add-datom e :resilience/chokepoint cp)
                 (add-datom e :resilience/chokepoint-load (get-in a [:choke-load cp]))
                 (add-datom e :resilience/cable-count (get-in a [:choke-count cp]))
                 (add-datom e :resilience/derived true)]))
            (sort-by (juxt #(- (get-in a [:choke-load %])) str) (keys (:choke-load a))))
    (mapcat (fn [s]
              (let [e (str "resilience-station-" s)]
                [(add-datom e :resilience/station s)
                 (add-datom e :resilience/station-degree (get-in a [:station-degree s]))
                 (add-datom e :resilience/station-capacity-tbps (get-in a [:station-capacity s]))
                 (add-datom e :resilience/derived true)]))
            (filter #(pos? (get-in a [:station-degree %]))
                    (sort-by (juxt #(- (get-in a [:station-degree %])) str) (keys stations))))
    (mapcat (fn [c]
              (let [e (str "resilience-cable-" c)]
                [(add-datom e :resilience/cable c)
                 (add-datom e :resilience/cable-diversity (get-in a [:cable-diversity c]))
                 (add-datom e :resilience/derived true)]))
            (sort-by (juxt #(get-in a [:cable-diversity %]) str) (keys cables)))
    (mapcat (fn [s]
              (let [e (str "resilience-gap-" s)]
                [(add-datom e :resilience/redundancy-gap-station s)
                 (add-datom e :resilience/station-degree (get-in a [:station-degree s]))
                 (add-datom e :resilience/derived true)]))
            (:redundancy-gap a)))))

;; ── content-addressed commit-DAG ──────────────────────────────────────────────
(defn- sha256-hex [^String s]
  (let [md (java.security.MessageDigest/getInstance "SHA-256")]
    (apply str (map #(format "%02x" (bit-and % 0xff)) (.digest md (.getBytes s "UTF-8"))))))

(defn- canonical [datoms prev]
  ;; deterministic canonical form (pr-str preserves vector order; map key order fixed here)
  (str "{:datoms " (pr-str datoms) " :prev " (pr-str prev) "}"))

(defn tx-cid
  "Content address = 'b' + sha256 over (prev, datoms) → a commit-DAG."
  ([datoms] (tx-cid datoms ""))
  ([datoms prev] (str "b" (sha256-hex (canonical datoms prev)))))

(defn make-tx [datoms & {:keys [tx-id as-of prev-cid] :or {prev-cid ""}}]
  {:tx/id tx-id :tx/as-of as-of :tx/prev prev-cid
   :tx/cid (tx-cid datoms prev-cid) :tx/count (count datoms) :tx/datoms datoms})

(defn append-tx
  "Append ONE transaction to the append-only log (never rewrites). Returns the tx CID."
  ([tx] (append-tx tx (log-default)))
  ([tx log-path]
   (let [f (io/file log-path)]
     (.mkdirs (.getParentFile (.getAbsoluteFile f)))
     (when-not (.exists f)
       (spit f (str ";; watatsuna kotoba Datom log — append-only EAVT transactions "
                    "(content-addressed DAG). Resilience map, never interdiction. "
                    "DO NOT hand-edit. ADR-2606012600.\n")))
     (spit f (str (pr-str tx) "\n") :append true)
     (:tx/cid tx))))

(defn read-log
  "Read the log back as a vector of transaction maps."
  ([] (read-log (log-default)))
  ([log-path]
   (let [f (io/file log-path)]
     (if-not (.exists f)
       []
       (->> (str/split-lines (slurp f))
            (map str/trim)
            (remove #(or (empty? %) (str/starts-with? % ";")))
            (mapv edn/read-string))))))

(defn head-cid
  ([] (head-cid (log-default)))
  ([log-path] (let [txs (read-log log-path)] (if (seq txs) (:tx/cid (last txs)) ""))))

(defn verify-chain
  "Recompute every CID from its datoms + prev; verify the DAG is intact. {:ok :length :broken-at}."
  ([] (verify-chain (log-default)))
  ([log-path]
   (let [txs (read-log log-path)]
     (loop [i 0 prev "" xs txs]
       (if (empty? xs)
         {:ok true :length (count txs) :broken-at -1}
         (let [tx (first xs)
               expect (tx-cid (:tx/datoms tx []) prev)]
           (if (or (not= (:tx/cid tx) expect) (not= (:tx/prev tx) prev))
             {:ok false :length (count txs) :broken-at i}
             (recur (inc i) (:tx/cid tx) (rest xs)))))))))
